### DISCUSSION TITLE
Can I call server.Send() inside my Events_DataReceived method?

### DISCUSSION THREAD

**Discussion by markregel2** [jchristn/SuperSimpleTcp#188] [Q&A]
I'm building a server that receives data from a client and forwards it to one or more other connected clients.  Is it appropriate to call Send from inside the Events_DataRecevied method? I have UseAsyncDataReceivedEvents = TRUE (the default), so my event is running in a thread. I had to make some minor modifications to Send to prevent exceptions if the client becomes disconnected (disposed) during Sending.  I'm new to c# Async programming, just looking for advice. Thanks.  Below is the simplified code:

SimpleTcpServer server;

//subscribe to events
server.Events.ClientConnected += Events_ClientConnected;
server.Events.ClientDisconnected += Events_ClientDisconnected;
server.Events.DataReceived += Events_DataReceived;

server.Start();

private void Events_DataReceived(object? sender, DataReceivedEventArgs e) 
{     
      server.Send("IpPort", e.Data.ToArray()); //is it a good idea to call send here?
}

**Comment by jchristn** [comment]
Hi @markregel2 you certainly can!  It really depends on how you want to structure your code and your application.  What I've seen work well is either a) directly calling from within the event or b) instantiating a separate class, from outside the event, that is accessible from within the event, which takes the responsibility of sending to a single or multiple (even broadcast) recipients.

**Reply by markregel2** [reply]
Hi @jchristn, thank you for this information, very helpful!  I'm instantiating a separate class for each client that connects (MyClient). These are maintained in a ConcurrentDictionary in the same way _clients are in SimpleTcpServer. When DataReceive fires, I'm calling MyClient.Receive() for the instance associated with the client. This method parses the incoming data looking for new packets. When a packet is found, it is forwarded to one or more (or all) of the other connected clients. One issue I'm having is my MyClient.Receive() method parses and assembles received data into packets. When data is received from a client, the method cannot be called again (from the same client) until it has finished processing the data.  One solution is to disable _settings.UseAsyncDataReceivedEvents. The other is to leave _settings.UseAsyncDataReceivedEvents enabled and use a semaphore (or lock) in my MyClient.Receive() method.  Any advice on which is better, pros and cons?  If I disable _settings.UseAsyncDataReceivedEvents, will it affect processing of other connected clients?
Thanks!

**Reply by jchristn** [reply]
I think I understand what you're saying - your ```MyClient.Receive()``` method is not thread-safe.  The SuperSimpleTcp library is itself thread-safe, meaning read and write operations will be serviced in order, but never trample on one another.  Personally in situations where some form of thread safety is required, I like to opt for a solution using locks/semaphores, only because a) the implementation is explicit and b) when reviewing the code later on it becomes clear why it was done.

**Reply by markregel2** [reply]
Hi @jchristn, The Server is working well for our project.  I took your advice and using semaphores to make my Receive() routine thread safe.  Our server is primarily a data communication switch. Numerous IoT devices connect to the server and communicate with other connected devices. Most of the data is very short packets (< 1000 bytes), but we can also send live audio. Our audio is very low bandwidth (100K bps). So far I've tested with 6 devices connected, and its working great with the live audio.  I need to do more testing at higher device counts (up to 1000), but do you have an idea of how the server will perform with 1000 devices, sending audio (100K bps) to all 1000 devices?  Assume the network bandwidth is able to support 1000 * 100K bps.  Thanks

**Reply by jchristn** [reply]
Hi @markregel2 that's great to hear!  I haven't done any performance or scale testing with this library and am excited to see how it works for you with that workload.  My only concern with this library (and any other TCP-library) is the cost of session setup/teardown (e.g. clients connecting and disconnecting).  That has always been the biggest problem area.  Please keep an eye on 1) thread count and 2) memory usage.  I don't expect there are memory leaks or tasks that aren't garbage collected, but those will be a leading indicator.

**Reply by markregel2** [reply]
Our clients connect once and remain permanently. The only time they reconnect is if they go offline  for some reason. Thanks for the advice, much appreciated.
